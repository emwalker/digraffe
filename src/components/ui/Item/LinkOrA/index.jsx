// @flow
import React from 'react'
import { Link } from 'found'
import isExternal from 'is-url-external'

import type { LocationDescriptor } from 'components/types'

type Props = {
  children: Iterable<React$Node> | string,
  className: string,
  to: LocationDescriptor,
}

const LinkOrA = ({ children, className, to }: Props) => {
  if (isExternal(to.pathname)) return <a href={to.pathname} className={className}>{ children }</a>
  return <Link to={to} className={className}>{ children }</Link>
}

export default LinkOrA
