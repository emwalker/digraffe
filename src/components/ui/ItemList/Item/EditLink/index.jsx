// @flow
import React, { Component } from 'react'
import { createFragmentContainer, graphql } from 'react-relay'

import Input from '../../Input'
import upsertLinkMutation from '../../../../../mutations/upsertLinkMutation'

type Props = {
  id: string,
  isOpen: boolean,
  relay: {
    environment: Object,
  },
  link: {
    resourceId: string,
    title: string,
    url: string,
  },
  toggleFn: Function,
}

type State = {
  title: string,
  url: string,
}

class EditLink extends Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      title: props.link.title,
      url: props.link.url,
    }
  }

  onSave = () => {
    const configs = []
    upsertLinkMutation(
      this.props.relay.environment,
      configs,
      {
        resourceId: this.props.link.resourceId,
        title: this.state.title,
        url: this.state.url,
      },
    )
    this.props.toggleFn()
  }

  render() {
    if (!this.props.isOpen)
      return null

    return (
      <div>
        <div className="d-flex col-12">
          <Input
            className="col-6"
            id={`edit-link-title-${this.props.id}`}
            label="Page title"
            value={this.state.title}
          />
          <Input
            className="col-6"
            id={`edit-link-url-${this.props.id}`}
            label="Url"
            value={this.state.url}
          />
        </div>
        <div>
          <button onClick={this.onSave} className="btn-primary">Save</button>
          {' '} or {' '}
          <button onClick={this.props.toggleFn} className="btn-link">cancel</button>
        </div>
      </div>
    )
  }
}

export default createFragmentContainer(EditLink, graphql`
  fragment EditLink_organization on Organization {
    resourceId
  }

  fragment EditLink_link on Link {
    resourceId
    title
    url
  }
`)
