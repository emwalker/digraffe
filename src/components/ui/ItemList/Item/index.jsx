// @flow
import React, { Component } from 'react'
import { pathOr } from 'ramda'
import classNames from 'classnames'

/* eslint no-underscore-dangle: 0 */

const edges = pathOr([], ['edges'])

type BadgeProps = {
  node: {
    name: string,
    resourcePath: string,
  }
}

const renderTopic = ({ node: { name, resourcePath } }: BadgeProps) => (
  <a className="Box-row-link" href={resourcePath} key={resourcePath}>
    <span
      className="Label swatch-green"
    >
      {name}
    </span>
  </a>
)

type Props = {
  __typename: string,
  display: string,
  resourcePath: string,
  topics: Array[],
}

class Item extends Component<Props> {
  get className(): string {
    return classNames(
      'Item-row',
      'Box-row',
      { 'Box-row--topic': this.props.__typename === 'Topic' },
    )
  }

  render() {
    return (
      <li
        className={this.className}
        key={this.props.resourcePath}
      >
        <div>
          <a className="Box-row-link" href={this.props.resourcePath}>
            { this.props.display || this.props.resourcePath }
          </a>
          { edges(this.props.topics).map(renderTopic) }
        </div>
        <div className="branch-name">{this.props.resourcePath}</div>
      </li>
    )
  }
}

export default Item
