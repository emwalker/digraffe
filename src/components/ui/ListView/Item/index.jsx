// @flow
import React from 'react'
import { pathOr } from 'ramda'

const edges = pathOr([], ['edges'])

type BadgeProps = {
  node: {
    name: string,
    resourcePath: string,
  }
}

const renderTopic = ({ node: { name, resourcePath } }: BadgeProps) => (
  <a className="Box-row-link" href={resourcePath} key={resourcePath}>
    <span
      className="Label swatch-green"
    >
      {name}
    </span>
  </a>
)

type Props = {
  display: string,
  resourcePath: string,
  topics: Array[],
}

export default ({ display, resourcePath, topics }: Props) => (
  <li
    className="Box-row"
    key={resourcePath}
  >
    <a className="Box-row-link" href={resourcePath}>
      { display }
    </a>
    { edges(topics).map(renderTopic) }
  </li>
)
