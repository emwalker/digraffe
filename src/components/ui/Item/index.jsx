// @flow
import React, { Component, Fragment, type Node } from 'react'
import classNames from 'classnames'

import type { TopicType } from 'components/types'
import LinkOrA from './LinkOrA'
import TopicBadge from '../TopicBadge'

/* eslint no-underscore-dangle: 0 */

type Props = {
  canEdit: boolean,
  children: Node,
  className: string,
  description?: ?string,
  displayColor: ?string,
  formIsOpen: boolean,
  newlyAdded: boolean,
  showEditButton: ?boolean,
  showLink: ?boolean,
  orgLogin: string,
  repoName: ?string,
  title: string,
  toggleForm: Function,
  topics: TopicType[],
  url: ?string,
}

class Item extends Component<Props> {
  static defaultProps = {
    description: null,
    displayColor: '#fff',
    showEditButton: false,
    showLink: true,
  }

  get className(): string {
    return classNames(
      'Item-row',
      'Box-row',
      this.props.className,
      { 'anim-fade-in': this.props.newlyAdded },
    )
  }

  get showEditButton(): boolean {
    return !this.props.formIsOpen && this.props.showEditButton === true
  }

  get style(): Object {
    if (!this.props.displayColor)
      return {}

    return {
      borderLeft: `5px solid ${this.props.displayColor}`,
    }
  }

  get url(): ?Node {
    if (!this.props.url || !this.props.showLink)
      return null

    return (
      <div
        className="mt-1 link-url branch-name css-truncate css-truncate-target"
      >
        { this.props.url }
      </div>
    )
  }

  get titleLink(): Node {
    if (!this.props.url) {
      return (
        // eslint-disable-next-line jsx-a11y/anchor-is-valid
        <a
          className="Box-row-link"
          href="#"
        >
          {this.props.title}
        </a>
      )
    }

    const to = this.locationDescriptor(this.props.url, this.props.title)

    return (
      <LinkOrA to={to} className="Box-row-link">
        { this.props.title }
      </LinkOrA>
    )
  }

  locationDescriptor = (pathname: string, itemTitle: string) => (
    {
      pathname,
      state: {
        orgLogin: this.props.orgLogin,
        repoName: this.props.repoName,
        itemTitle,
      },
    }
  )

  renderTopicBadge = (
    // eslint-disable-next-line react/no-unused-prop-types
    { name, resourcePath }: {name: string, resourcePath: string},
  ) => (
    <TopicBadge
      key={resourcePath}
      name={name}
      to={this.locationDescriptor(resourcePath, name)}
    />
  )

  renderEditable = () => (
    <Fragment>
      <div className="clearfix d-flex flex-items-center">
        <div className="col-10">
          <div>
            { this.titleLink }
            <div>{ this.props.description }</div>
          </div>
          { this.url }
          <div>
            { this.props.topics.map(this.renderTopicBadge) }
          </div>
        </div>
        <div className="col-2 text-center">
          { this.showEditButton &&
            <button onClick={this.props.toggleForm} className="btn-link">Edit</button>
          }
        </div>
      </div>
      <div>
        { this.props.formIsOpen && this.props.children }
      </div>
    </Fragment>
  )

  renderWide = () => (
    <div className="clearfix d-flex flex-items-center">
      <div className="col-12">
        <div>
          { this.titleLink }
          <div>{ this.props.description }</div>
        </div>
        { this.url }
        <div>
          { this.props.topics.map(this.renderTopicBadge) }
        </div>
      </div>
    </div>
  )

  render = () => {
    const { url } = this.props

    return (
      <li
        className={this.className}
        style={this.style}
        key={url}
      >
        { this.props.canEdit
          ? this.renderEditable()
          : this.renderWide()
        }
      </li>
    )
  }
}

export default Item
