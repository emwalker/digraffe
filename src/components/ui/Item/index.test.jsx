import React from 'react'
import { shallow } from 'enzyme'

import Item from './index'

const props = {
  children: 'child',
  className: '',
  formIsOpen: false,
  newlyAdded: false,
  loading: false,
  title: 'title',
  toggleForm: jest.fn(),
  topics: [],
  url: null,
}

describe('<Item />', () => {
  const wrapper = shallow(<Item {...props} />)

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('loading', () => {
    const editLink = () => wrapper.find('button.btn-link')

    describe('when loading: true', () => {
      beforeEach(() => {
        wrapper.setProps({ loading: true })
      })

      it('hides the edit link', () => {
        expect(editLink().exists()).toBeFalsy()
      })
    })

    describe('when loading: false', () => {
      beforeEach(() => {
        wrapper.setProps({ loading: false })
      })

      it('hides the edit link', () => {
        expect(editLink().exists()).toBeTruthy()
      })
    })
  })
})
