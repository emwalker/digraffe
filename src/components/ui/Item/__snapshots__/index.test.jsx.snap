// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item /> renders 1`] = `
<li
  className="Item-row Box-row"
  key="null"
  style={
    Object {
      "borderLeft": "5px solid #fff",
    }
  }
>
  <div
    className="clearfix d-flex flex-items-center"
  >
    <div
      className="col-10"
    >
      <div>
        <a
          className="Box-row-link"
          href="#"
        >
          title
        </a>
        <div />
      </div>
      <div />
    </div>
    <div
      className="col-2 text-center"
    />
  </div>
  <div />
</li>
`;
